pkg/types/intoto/v0.0.3/entry.go

```diff
@@ -0,0 +1,460 @@
+		result = append(result, payloadHashKey)
+	}
+
+	if v.IntotoObj.EnvelopeHash != nil {
+		envelopeHashKey := strings.ToLower(fmt.Sprintf("%s:%s", *v.IntotoObj.EnvelopeHash.Algorithm, *v.IntotoObj.EnvelopeHash.Value))
+		result = append(result, envelopeHashKey)
+	}
+
+	if v.env == nil {
+		log.Logger.Info("IntotoObj content or DSSE envelope is nil, returning partial set of keys")
+		return result, nil
+	}
+
+	switch v.env.PayloadType {
+	case in_toto.PayloadType:
+
+		if v.env.Payload == "" {
+			log.Logger.Info("IntotoObj DSSE payload is empty")
+			return result, nil
+		}
+		decodedPayload, err := v.env.DecodeB64Payload()
+		if err != nil {
+			return result, fmt.Errorf("could not decode envelope payload: %w", err)
+		}
+		statement, err := parseStatement(decodedPayload)
+		if err != nil {
+			return result, err
+		}
+		for _, s := range statement.Subject {
+			for alg, ds := range s.Digest {
+				result = append(result, alg+":"+ds)
+			}
+		}
+		// Not all in-toto statements will contain a SLSA provenance predicate.
+		// See https://github.com/in-toto/attestation/blob/main/spec/README.md#predicate
+		// for other predicates.
+		if predicate, err := parseSlsaPredicate(decodedPayload); err == nil {
+			if predicate.Predicate.Materials != nil {
+				for _, s := range predicate.Predicate.Materials {
+					for alg, ds := range s.Digest {
+						result = append(result, alg+":"+ds)
+					}
+				}
+			}
+		}
+	default:
+		log.Logger.Infof("Unknown in_toto DSSE envelope Type: %s", v.env.PayloadType)
+	}
+	return result, nil
+}
+
+func parseStatement(p []byte) (*in_toto.Statement, error) {
+	ps := in_toto.Statement{}
+	if err := json.Unmarshal(p, &ps); err != nil {
+		return nil, err
+	}
+	return &ps, nil
+}
+
+func parseSlsaPredicate(p []byte) (*in_toto.ProvenanceStatement, error) {
+	predicate := in_toto.ProvenanceStatement{}
+	if err := json.Unmarshal(p, &predicate); err != nil {
+		return nil, err
+	}
+	return &predicate, nil
+}
+
+func (v *V003Entry) Unmarshal(pe models.ProposedEntry) error {
+	it, ok := pe.(*models.Intoto)
+	if !ok {
+		return errors.New("cannot unmarshal non Intoto v0.0.3 type")
+	}
+
+	intotoObj := &models.IntotoV003Schema{}
+
+	if err := types.DecodeEntry(it.Spec, intotoObj); err != nil {
+		return err
+	}
+
+	// field validation
+	if err := intotoObj.Validate(strfmt.Default); err != nil {
+		return err
+	}
+
+	// either we have just proposed content or the canonicalized fields
+	if intotoObj.ProposedContent == nil {
+		// then we need canonicalized fields, and all must be present (if present, they would have been validated in the above call to Validate())
+		if intotoObj.EnvelopeHash == nil || intotoObj.PayloadHash == nil || len(intotoObj.Signatures) == 0 {
+			return errors.New("either proposed content or envelope hash, payload hash, and signatures must be present")
+		}
+		v.IntotoObj = *intotoObj
+		return nil
```

Comment:
Because this is incoming proposed content with canonicalized fields, shouldn't `verifyEnvelope` still apply here for robustness?

What if a user proposed a `ProposedEntry` with pre-canonicalized fields, let's say, because it wants to hide the `Envelope` content? It should still be verifiable, assuming verification over a hash.

But if it's clear that users MUST be publishing full envelope content to the log, fine with me.

Reply:
Only hesitation here is that I am wondering if it's possible for a client to be setting the `EnvelopeHash`, `PayloadHash`, and `Signatures` of a bogus signed envelope and bypassing validation (not through the CLI, but through setting a proposed entry in the createentry handler)